File: 백준/Gold/13418. 학교 탐방하기/학교 탐방하기.py
n, m = map(int, input().split())
edges = []

for _ in range(m+1):
    a, b, type = map(int, input().split())
    edges.append((type, a, b))

def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]

def union_parent(a, b, parent):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

def kruskal(MAP, sorte_type):
    parent = [i for i in range(n+1)]
    count = 0
    sorted_edge = sorted(MAP, reverse=sorte_type)
    for t, a, b in sorted_edge:
        if find_parent(parent, a) != find_parent(parent, b):
            union_parent(a, b, parent)
            if t == 0: #오르막 길이면 count 증가
                count += 1
    return count

max_tired = kruskal(edges, False) #최대 피로도 계산(오르막 순으로 정렬 후 union-find
min_tired = kruskal(edges, True) #최소 피로도 계산 (내리막 순으로 정렬 후 unoin-find)

print(max_tired ** 2 - min_tired ** 2)

File: 백준/Gold/13418. 학교 탐방하기/학교 탐방하기.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] input = br.readLine().split(" ");
		int n = Integer.parseInt(input[0]);
		int m = Integer.parseInt(input[1]);
		int[] parent = new int[n+1];
		for(int i=0; i<n+1; i++) {
			parent[i] = i;
		}

		Edge[] edges = new Edge[m+1];
		for(int i=0;i<m+1;i++) {
			String[] info = br.readLine().split(" ");
			int a = Integer.parseInt(info[0]);
			int b = Integer.parseInt(info[1]);
			int c = Integer.parseInt(info[2]);
			edges[i] = new Edge(a, b, c);
		}
		Arrays.sort(edges, (a, b) -> a.distance - b.distance); //오름차순; 오르막이 0 -> 최대신장트리
		
		int max = 0;
		for (int i=0; i<edges.length; i++) {
			Edge e = edges[i];
			if(findParent(e.a, parent) != findParent(e.b, parent)) {
				union(e.a, e.b, parent);
				if (e.distance == 0){
					max++;
				}
			}
		}

		for(int i=0; i<n+1; i++) {
			parent[i] = i;
		}

		int min = 0;
		for (int i=edges.length-1; i>=0; i--) {
			Edge e = edges[i];
			if(findParent(e.a, parent) != findParent(e.b, parent)) {
				union(e.a, e.b, parent);
				if (e.distance == 0){
					min++;
				}
			}
		}
		
		System.out.println(max * max - min * min);

	}
	private static int findParent(int x, int[] parent) {
		if(parent[x] != x) {
			parent[x] = findParent(parent[x], parent);
		}
		return parent[x];
	}

	private static void union(int a, int b, int[] parent) {
		a = findParent(a, parent);
		b = findParent(b, parent);
		if(a < b) {
			parent[b] =a;
		} else {
			parent[a] = b;
		}
	}
}
class Edge {
	int a, b, distance;

	Edge(int a, int b, int distance) {
		this.a = a;
		this.b = b;
		this.distance = distance;
	}
}
